src/python/grpcio_testing/grpc_testing/_channel.py

```diff
@@ -0,0 +1,753 @@
+
+
+def _state_cancel(state, code, details):
+    with state.condition:
+        if state.code is None:
+            if state.initial_metadata is None:
+                state.initial_metadata = _common.FUSSED_EMPTY_METADATA
+            state.trailing_metadata = _common.FUSSED_EMPTY_METADATA
+            state.code = code
+            state.details = details
+            state.condition.notify_all()
+            return True
+        else:
+            return False
+
+
+def _state_terminate(state):
+    with state.condition:
+        while True:
+            if state.code is None:
+                state.condition.wait()
+            else:
+                return state.trailing_metadata, state.code, state.details
+
+
+def _state_is_active(state):
+    with state.condition:
+        return state.code is None
+
+
+def _state_time_remaining(unused_state):
+    raise NotImplementedError()
+
+
+def _state_add_callback(unused_state, unused_callback):
+    raise NotImplementedError()
+
+
+def _state_initial_metadata(state):
+    with state.condition:
+        while True:
+            if state.initial_metadata is None:
+                if state.code is None:
+                    state.condition.wait()
+                else:
+                    return _common.FUSSED_EMPTY_METADATA
+            else:
+                return state.initial_metadata
+
+
+def _state_trailing_metadata(state):
+    with state.condition:
+        while state.trailing_metadata is None:
+            state.condition.wait()
+        return state.trailing_metadata
+
+
+def _state_code(state):
+    with state.condition:
+        while state.code is None:
+            state.condition.wait()
+        return state.code
+
+
+def _state_details(state):
+    with state.condition:
+        while state.details is None:
+            state.condition.wait()
+        return state.details
+
+
+def _handler_cancel(handler):
+    return handler.cancel(grpc.StatusCode.CANCELLED, 'Locally cancelled!')
+
+
+def _handler_is_active(handler):
+    return handler.is_active()
+
+
+def _handler_time_remaining(unused_handler):
+    raise NotImplementedError()
+
+
+def _handler_add_callback(handler, callback):
+    return handler.add_callback(callback)
+
+
+def _handler_initial_metadata(handler):
+    return handler.initial_metadata()
+
+
+def _handler_trailing_metadata(handler):
+    trailing_metadata, unused_code, unused_details = handler.terminate()
+    return trailing_metadata
+
+
+def _handler_code(handler):
+    unused_trailing_metadata, code, unused_details = handler.terminate()
+    return code
+
+
+def _handler_details(handler):
+    unused_trailing_metadata, unused_code, details = handler.terminate()
+    return details
+
+
+class _Call(grpc.Call):
+
+    def __init__(self, handler):
+        self._handler = handler
+
+    def cancel(self):
+        _handler_cancel(self._handler)
+
+    def is_active(self):
+        return _handler_is_active(self._handler)
```

Comment:
There's kind of a long story here that @kpayson64 already knows, and it starts with [the _Rendezvous class in the current gRPC Python](https://github.com/grpc/grpc/blob/71a78e92cb9f6af405e6f80c9a31c408c4025095/src/python/grpcio/grpc/_channel.py#L238). That's a class that implements a bunch of interfaces, but at each point where an instance of the class is passed from gRPC Python to the application the instance that is passed is only valid for use according to some subset of those interfaces. Authors have introspected the instance, seen a lot of methods that look legitimate, and dug themselves into holes trying crazy things. The motivation in this module for pulling all the behavior out into module-scope functions is so that we can have these several classes that each only implement the combinations of interfaces that are needed to be passed, and the indirection is so that exactly all logic is done in shared module-scope functions. It's only a little indirection, so that's not an issue, and it's a lot of text on the page, but it's module-private implementation inside package-private module, so the code golfing of using dynamic magic to shrink it down textually can wait until later.

I guess the short answer is "this prototypes something I've been wanting to try in the non-test codebase for a while".